nix: Fix the macOS build (nix package + devShell)

rrbutani opened the pull request:
Three changes to unbreak the macOS build of the nix package for `zed-editor` as well as building zed inside of the nix devShell:
  - updating `rust-overlay` to a version w/1.86
  - updating `livekit-libwebrtc` as a stopgap until NixOS/nixpkgs#396016 is merged (h/t: @WeetHet)
  - dropping `-all_load` to fix linker errors when using `ld64` compatible linkers that aren't `ld_prime`

The commits within have more details but I think the last one is the only interesting/contentious change so I've pasted its description below.

I've also got some other improvements/"fixes" that I'll open followup PRs about but I believe this PR should be enough to fix the macOS nix build 🤞. Planned follow ups:
  - having nix devShell builds use `ld` on the host system if present and suggesting that users install it in `shellHook`
     + @P1n3appl3's testing showed that when linking dev builds of `zed-editor`, `ld_prime` is significantly faster (~1.2s) than the open source `ld64` (~4s) as well as `lld` (~3.2s)
  - having the devShell hook tell users that don't have an Xcode installation to pass `--features gpui/runtime_shaders` or to install Xcode

---

from 01af0a639ab738062c56346eac67d29cd3a518e3:

With the `-all_load` linker flag, we get errors like this when linking the zed-editor binary (both in the nix devShell and when building the zed-editor nix package):
```plain
duplicate symbol '___aarch64_have_lse_atomics' in:
    /nix/store/mg65rms59mr0jwn3pg0qzvsn8r27cpb6-rust-std-1.85.1-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-a94b6215175cddb2.rlib(ad3ac4dcdcbf93cb-aarch64.o)
    /nix/store/5qawh29l2yhzj0l9a7ybbb3nzwkpis4a-clang-wrapper-19.1.7/resource-root/lib/darwin/libclang_rt.osx.a(aarch64.c.o)
duplicate symbol '__aarch64_cas1_relax' in:
    /nix/store/mg65rms59mr0jwn3pg0qzvsn8r27cpb6-rust-std-1.85.1-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-a94b6215175cddb2.rlib(1dfec86f22ed8170-lse_cas1_relax.o)
    /nix/store/5qawh29l2yhzj0l9a7ybbb3nzwkpis4a-clang-wrapper-19.1.7/resource-root/lib/darwin/libclang_rt.osx.a(outline_atomic_cas1_1.S.o)
duplicate symbol '__aarch64_cas1_acq' in:
    /nix/store/mg65rms59mr0jwn3pg0qzvsn8r27cpb6-rust-std-1.85.1-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-a94b6215175cddb2.rlib(1dfec86f22ed8170-lse_cas1_acq.o)
    /nix/store/5qawh29l2yhzj0l9a7ybbb3nzwkpis4a-clang-wrapper-19.1.7/resource-root/lib/darwin/libclang_rt.osx.a(outline_atomic_cas1_2.S.o)
<snipped...>
duplicate symbol '__aarch64_ldset8_acq_rel' in:
    /nix/store/mg65rms59mr0jwn3pg0qzvsn8r27cpb6-rust-std-1.85.1-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-a94b6215175cddb2.rlib(1dfec86f22ed8170-lse_ldset8_acq_rel.o)
    /nix/store/5qawh29l2yhzj0l9a7ybbb3nzwkpis4a-clang-wrapper-19.1.7/resource-root/lib/darwin/libclang_rt.osx.a(outline_atomic_ldset8_4.S.o)
ld: 148 duplicate symbols for architecture arm64
```

> [!TIP]
> To get the final linker invocation (such that you can run it yourself) use:
> ```bash
> RUSTC_LOG=rustc_codegen_ssa::back::link=info cargo rustc -p zed -- \
>   -C save-temps=true \
>   -C link-arg=-save-temps \
>   -C link-arg=-v
> ```
>
> `-C save-temps` instructs `rustc` to persist its temporary files (i.e. intermediary `.o` files not in `$TARGET_DIR`)
>
> `-C link-arg=-save-temps` instructs the linker driver (clang) to persist its temporary files (i.e. response file fed to the linker)
>
> `-C link-arg=v` tells the linker driver (clang) to show us the linker invocation
>
> `RUSTC_LOG=rustc_codegen_ssa::back::link=info` is how we get rustc to show us stdout/stderr from the linker driver (better ways to do this are not yet in stable rustc IIUC)

Inspecting the final linker invocation reveals that the linker is being passed both:
  - `<rust install>/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-<hash>.rlib`
  - `-lclang_rt.osx`

`-lclang_rt.osx` resolves to `libclang_rt.osx.a` within clang's resource root
  - when using a toolchain from nixpkgs this lives at the path in the error message above
  -  when using a toolchain from Xcode this comes from within the Xcode toolchains directory
     + look within the path `clang -print-resource-dir` prints
     + or run something like `clang -v -xc - <<<"int main() {}"` and inspect the resulting linker invocation's args to find it, should be one of the last args

Comparing `libclang_rt.osx.a` from the Xcode toolchain and from nixpkgs does not yield any answers; both have these atomic intrinsics as regular symbols in the text section (not weak).

`libcompiler_builtins.rlib` is also not at fault here – the nix-supplied version (via `rust-overlay`) does actually come from rust-lang.org and is identical to what's used outside of nix. This `rlib` is meant to have these symbols (IIUC, because LLVM may emit calls to these intrinsics during codegen) and `rustc` passes this rlib to the linker along with `libstd` and its other dependencies.

The issue is the combination of the `-all_load` and `-lclang_rt.osx` flags: `-all_load` instructs the linker to use all symbols from static archives passed to the linker and `-lclang_rt.osx` asks the linker to include `libclang_rt.osx.a`.

Both of these are unusual:
  - `-all_load` is specified for the zed build by `.cargo/config.toml`
  - `-lclang_rt.osx` is specified `webrtc-sys`'s `build.rs` here:
    + https://github.com/livekit/rust-sdks/blob/rust-sdks/webrtc-sys%400.3.7/webrtc-sys/build.rs#L253-L260

Both of these flags are required to hit this issue.

`rustc` passes `-nodefaultlibs` to the linker driver (`clang`) which disables its default behavior of passing `libclangrt.osx.a` to the linker – if not for `webrtc-sys`, `libclang_rt.osx.a` would not be linked in and `-all_load` would not result in us having duplicate symbol issues.

> [!WARNING]
> So why did folks using an Xcode provided toolchain not run into this issue?
>
> It appears that this is a bug in `ld_prime` (the closed-source `ld64` rewrite that's part of Apple's `cctools`).
>
> `-all_load` + `libcompiler_builtins-<hash>.rlib` + `libclang_rt.osx.a` _should_ (AIUI) result in duplicate symbol errors and when using other `ld64`-compatible linkers it does; all of the following yield the duplicate symbol error:
>   - modern `ld` from Xcode with the `-ld_classic` flag
>   - `ld` from cctools in nixpkgs (built from https://github.com/apple-oss-distributions/cctools)
>   - `lld` (from LLVM; tested v19.1.7)
>
> Unfortunately since `ld_prime` is not open source it's difficult to confirm that this really is a bug and not purposeful but passing `-why_load` to the linker provides some insight into the linker's thought process.
>
> `ld_prime` says that `libclang_rt.osx.a` is pulled in via another symbol (`chk_stack` IIRC) that a previous item on the command line required; perhaps this causes `ld_prime` to (incorrectly) use its normal (not `-all_load`) behavior for that one static archive...

To "fix" this issue we have two options: remove `-all_load` or remove `-lclang_rt.osx`.

On advice from @P1n3appl3, this PR does the former: it's not clear whether `-all_load` is still required in the build and this is also the easier option (doesn't require patching `webrtc-sys`).

---

Release Notes:

- N/A

rrbutani committed:
nix: move `--cfg gles` to the main zed-editor derivation

(this was incorrectly placed in the `webrtc-sys` crate's derivation as
part of #27881)

rrbutani committed:
nix: update flake inputs

in particular: need to bump `rust-overlay` so rust 1.86 is available
(to match #28021)

rrbutani committed:
nix: temporarily patch `livekit-libwebrtc` to m125

(required by #27126; `libwebrtc-sys` fails to build against `m114`)

See: NixOS/nixpkgs#396016

`webrtc-sources.nix` is courtesy of @WeetHet

Co-authored-by: Stanislav Alekseev <43210583+WeetHet@users.noreply.github.com>

rrbutani committed:
build: drop `-all_load` on macOS

With the `-all_load` linker flag, we get errors like this when linking
the zed-editor binary (both in the nix devShell and when building the
zed-editor nix package):
```plain
duplicate symbol '___aarch64_have_lse_atomics' in:
    /nix/store/mg65rms59mr0jwn3pg0qzvsn8r27cpb6-rust-std-1.85.1-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-a94b6215175cddb2.rlib(ad3ac4dcdcbf93cb-aarch64.o)
    /nix/store/5qawh29l2yhzj0l9a7ybbb3nzwkpis4a-clang-wrapper-19.1.7/resource-root/lib/darwin/libclang_rt.osx.a(aarch64.c.o)
duplicate symbol '__aarch64_cas1_relax' in:
    /nix/store/mg65rms59mr0jwn3pg0qzvsn8r27cpb6-rust-std-1.85.1-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-a94b6215175cddb2.rlib(1dfec86f22ed8170-lse_cas1_relax.o)
    /nix/store/5qawh29l2yhzj0l9a7ybbb3nzwkpis4a-clang-wrapper-19.1.7/resource-root/lib/darwin/libclang_rt.osx.a(outline_atomic_cas1_1.S.o)
duplicate symbol '__aarch64_cas1_acq' in:
    /nix/store/mg65rms59mr0jwn3pg0qzvsn8r27cpb6-rust-std-1.85.1-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-a94b6215175cddb2.rlib(1dfec86f22ed8170-lse_cas1_acq.o)
    /nix/store/5qawh29l2yhzj0l9a7ybbb3nzwkpis4a-clang-wrapper-19.1.7/resource-root/lib/darwin/libclang_rt.osx.a(outline_atomic_cas1_2.S.o)
<snipped...>
duplicate symbol '__aarch64_ldset8_acq_rel' in:
    /nix/store/mg65rms59mr0jwn3pg0qzvsn8r27cpb6-rust-std-1.85.1-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-a94b6215175cddb2.rlib(1dfec86f22ed8170-lse_ldset8_acq_rel.o)
    /nix/store/5qawh29l2yhzj0l9a7ybbb3nzwkpis4a-clang-wrapper-19.1.7/resource-root/lib/darwin/libclang_rt.osx.a(outline_atomic_ldset8_4.S.o)
ld: 148 duplicate symbols for architecture arm64
```

> [!TIP]
> To get the final linker invocation (such that you can run it yourself)
> use:
> ```bash
> RUSTC_LOG=rustc_codegen_ssa::back::link=info cargo rustc -p zed -- \
>   -C save-temps=true \
>   -C link-arg=-save-temps \
>   -C link-arg=-v
> ```
>
> `-C save-temps` instructs `rustc` to persist its temporary files (i.e.
> intermediary `.o` files not in `$TARGET_DIR`)
>
> `-C link-arg=-save-temps` instructs the linker driver (clang) to
> persist its temporary files (i.e. response file fed to the linker)
>
> `-C link-arg=v` tells the linker driver (clang) to show us the linker
> invocation
>
> `RUSTC_LOG=rustc_codegen_ssa::back::link=info` is how we get rustc to
> show us stdout/stderr from the linker driver (better ways to do this
> are not yet in stable rustc IIUC)

Inspecting the final linker invocation reveals that the linker is being
passed both:
  - `<rust install>/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-<hash>.rlib`
  - `-lclang_rt.osx`

`-lclang_rt.osx` resolves to `libclang_rt.osx.a` within clang's resource
root – when using a toolchain from nixpkgs this lives at the path in the
error message above; when using a toolchain from Xcode this comes from
within the Xcode toolchains directory (look within the path
`clang -print-resource-dir` prints or run something like
`clang -v -xc - <<<"int main() {}"` and inspect the resulting linker
invocation's args to find it, should be one of the last args)

Comparing `libclang_rt.osx.a` from the Xcode toolchain and from nixpkgs
does not yield any answers; both have these atomic intrinsics as regular
symbols in the text section (not weak).

`libcompiler_builtins.rlib` is also not at fault here – the nix-supplied
version (via `rust-overlay`) does actually come from rust-lang.org and
is identical to what's used outside of nix. This `rlib` is meant to have
these symbols (IIUC, because LLVM may emit calls to these intrinsics
during codegen) and `rustc` passes this rlib to the linker along with
`libstd` and its other dependencies.

The issue is the combination of the `-all_load` and `-lclang_rt.osx`
flags: `-all_load` instructs the linker to use all symbols from static
archives passed to the linker and `-lclang_rt.osx` asks the linker to
include `libclang_rt.osx.a`.

Both of these are unusual:
  - `-all_load` is specified for the zed build by `.cargo/config.toml`
  - `-lclang_rt.osx` is specified `webrtc-sys`'s `build.rs` here:
    + https://github.com/livekit/rust-sdks/blob/rust-sdks/webrtc-sys%400.3.7/webrtc-sys/build.rs#L253-L260

Both of these flags are required to hit this issue.

`rustc` passes `-nodefaultlibs` to the linker driver (`clang`) which
disables its default behavior of passing `libclangrt.osx.a` to the
linker – if not for `webrtc-sys`, `libclang_rt.osx.a` would not be
linked in and `-all_load` would not result in us having duplicate symbol
issues.

> [!WARNING]
> So why did folks using an Xcode provided toolchain not run into this
> issue?
>
> It appears that this is a bug in `ld_prime` (the closed-source `ld64`
> rewrite that's part of Apple's `cctools`).
>
> `-all_load` + `libcompiler_builtins-<hash>.rlib` + `libclang_rt.osx.a`
> _should_ (AIUI) result in duplicate symbol errors and when using other
> `ld64`-compatible linkers it does; all of the following yield the
> duplicate symbol error:
>   - modern `ld` from Xcode with the `-ld_classic` flag
>   - `ld` from cctools in nixpkgs (built from
>     https://github.com/apple-oss-distributions/cctools)
>   - `lld` (from LLVM; tested v19.1.7)
>
> Unfortunately since `ld_prime` is not open source it's difficult to
> confirm that this really is a bug and not purposeful but passing
> `-why_load` to the linker provides some insight into the linker's
> thought process: `ld_prime` says that `libclang_rt.osx.a` is pulled
> in via another symbol (`chk_stack` IIRC) that a previous item on the
> command line required; perhaps this causes `ld_prime` to (incorrectly)
> use its normal (not `-all_load`) behavior for that one static
> archive...

---

To "fix" this issue we have two options: remove `-all_load` or remove
`-lclang_rt.osx`.

On advice from @P1n3appl3, this commit does the former: it's not clear
whether `-all_load` is still required in the build and this is also the
easier option (doesn't require patching `webrtc-sys`).

Co-authored-by: Julia Ryan <juliaryan3.14@gmail.com>

mikayla-maki commented:
@rrbutani The test failure is caused by removing that `-all_load`. I'm no expert in building code, but we use `[linkme](https://github.com/dtolnay/linkme)` to make distribute action registration across the codebase. Code we do something that's more targeted, and only affects the Zed crates?

P1n3appl3 commented:
I think we could use the `-force_load <archive>` flag to only have that behavior for the specific crates we need? That way we wouldn't run into the error with clang_rt and compilerbuiltins.

WeetHet commented:
Your livekit-libwebrtc is missing the `pkgs/by-name/li/livekit-libwebrtc/0003-fix-boringssl-generic.patch` patch. Without it, webrtc doesn't build on Linux because GCC 14 doesn't support `_Generic` in C++ mode

rrbutani commented:
> Without it, webrtc doesn't build on Linux because GCC 14 doesn't support `_Generic` in C++ mode

thanks, had not tested linux yet; will address

rrbutani committed:
nix: add `xorg.libX11` to deps for linux

required since #27807

rrbutani committed:
nix: fixes and tweaks for CI

specifically:
  - `-L` to print build logs
  - `--accept-flake-config` to use the cachix substituter in CI
  - add `|| :` so that the nix store cleanup step doesn't fail when
    cleanup isn't required

SomeoneToIgnore commented:
The build seems fixed at this moment of time, given that https://github.com/zed-industries/zed/pull/33688 exists.
I will close this one as stale for now, as there are conflicts and tests failing.
